### ISSUE TITLE
Question: Shared Cache for Windows, Linux and OSX

### ISSUE THREAD

**Issue by missmah** [fastbuild/fastbuild#743]
Hi @ffulin,

This issue is just a question about optimization of shared caches.

Are there any pros/cons that would make it clear whether shared caches should be shared by multiple separate OSes (e.g. `Windows`, `Linux` and `OSX`), or whether each should get their own cache?

That is, one `.fbuild.cache` shared by all, or three `.fbuild.cache.win32`, `.fbuild.cache.linux`, `.fbuild.cache.osx`.

I assume the cache lookups get slightly slower and the hash tables a bit bigger in memory as the cache grows. This leads me to think that separating out platforms which won't be able to hit on each other's cache entries is probably a win.

OTOH, it means three separate clear cache nightly operations (not a big deal) and three separate cache size limits on disk (might have its own pros and cons list).

Any feedback you have would be appreciated.

Thanks & Best,
Layla

**Comment by ffulin** [comment]
**TL;DR:**

I recommend, and personally use, a single global cache. It's easier to configure and manage. Other factors are not a concern at all.

If you do any tests and find anything that might counter that, I'd certainly like to know about it.

**A more detailed answer:**

As far as the performance or functionality of FASTBuild is concerned it makes little difference whether all your targets share the same cache path. The cache is partitioned in a hierarchy specifically designed to ensure it scales well, even to hundreds of millions of entries.

There is some reduction in hash collision potential gained by splitting the cache, but the chance of such a collision is already so infinitesimally small that it's essentially a non-issue. Additionally, the difference in the toolchain binaries contribution to the cache key is such that they are effectively already namespaced.

**> I assume the cache lookups get slightly slower**

The cache file system is structured in a hierarchy to ensure no given level contains very many entries.  No directory in the cache is likely to have anywhere near the number of files in it that you'd even need to worry about scalability. It is specifically designed to work even with poorly performing filesystems

As long as you're not hosting your cache on FAT32 with Windows short-name generation enabled, you're probably fine :)

**> and the hash tables a bit bigger in memory as the cache grows.** 

There is no hash table or in-memory structure of any kind. The filesystem is the database. All read/write operations operate directly on one file with no knowledge of anything else present in the cache.

**> three separate cache size limits on disk**

If isolating disk use quotas between platforms is important for some reason, splitting the cache would allow for that. I can't think of a use for that personally, short of some kind of accounting or cost-attibution reason, but even then that seems unlikely to be useful.

**Comment by ffulin** [comment]
I'm going to close this issue as answered. Please let me know if you have any follow up questions.
